regression-tests.dnsdist/test_Trailing.py

```diff
@@ -40,35 +87,257 @@ def testTrailingAllowed(self):
+        expectedResponse = dns.message.make_response(query)
+        expectedResponse.set_rcode(dns.rcode.SERVFAIL)
+
+        for method in ("sendUDPQuery", "sendTCPQuery"):
+            sender = getattr(self, method)
+            # (receivedQuery, receivedResponse) = self.sendUDPQuery(query, response)
+            # (receivedQuery, receivedResponse) = self.sendTCPQuery(query, response)
+            (_, receivedResponse) = sender(query, response)
+            self.assertTrue(receivedResponse)
+            self.assertEquals(receivedResponse, expectedResponse)
+
+    def testTrailingAdded(self):
+        """
+        Trailing data: Add
+
+        """
+        name = 'added.trailing.tests.powerdns.com.'
+        query = dns.message.make_query(name, 'A', 'IN')
+        response = dns.message.make_response(query)
+        rrset = dns.rrset.from_text(name,
+                                    3600,
+                                    dns.rdataclass.IN,
+                                    dns.rdatatype.A,
+                                    '127.0.0.1')
+        response.answer.append(rrset)
+        expectedResponse = dns.message.make_response(query)
+        expectedResponse.set_rcode(dns.rcode.REFUSED)
+
+        for method in ("sendUDPQuery", "sendTCPQuery"):
+            sender = getattr(self, method)
+            # (receivedQuery, receivedResponse) = self.sendUDPQuery(query, response)
+            # (receivedQuery, receivedResponse) = self.sendTCPQuery(query, response)
+            (receivedQuery, receivedResponse) = sender(query, response)
+            self.assertTrue(receivedQuery)
+            self.assertTrue(receivedResponse)
+            receivedQuery.id = query.id
+            self.assertEquals(receivedQuery, query)
+            self.assertEquals(receivedResponse, expectedResponse)
+
+class TestTrailingDataToDnsdist(DNSDistTest):
+    _config_template = """
+    newServer{address="127.0.0.1:%s"}
+
+    addAction(AndRule({QNameRule("dropped.trailing.tests.powerdns.com."), TrailingDataRule()}), DropAction())
+
+    function removeTrailingData(dq)
+        local success = dq:setTrailingData({})
+        if not success then
+            return DNSAction.ServFail, ""
+        end
+        return DNSAction.None, ""
+    end
+    addLuaAction("removed.trailing.tests.powerdns.com.", removeTrailingData)
+
+    function reportTrailingData(dq)
+        local tailBytes = dq:getTrailingData()
+        local tailChars = string.char(unpack(tailBytes))
+        return DNSAction.Spoof, "-" .. tailChars .. ".echoed.trailing.tests.powerdns.com."
+    end
+    addLuaAction("echoed.trailing.tests.powerdns.com.", reportTrailingData)
+
+    function replaceTrailingData(dq)
+        local success = dq:setTrailingData({65, 66, 67}) -- "ABC"
+        if not success then
+            return DNSAction.ServFail, ""
+        end
+        return DNSAction.None, ""
+    end
+    addLuaAction("replaced.trailing.tests.powerdns.com.", replaceTrailingData)
```

Comment:
This line looks like a c/p mistake.

Reply:
No, the idea is that queries for replaced.trailing.tests.powerdns.com will be subject first to `replaceTrailingData` (which sets the data to ASCII "ABC") and then to `reportTrailingData` (which uses trailing data to spoof a response), allowing us to verify that the first succeeded.

Reply:
ah, right, I missed that! Thanks for the explanation!